internal/shakex4/f1600x4_amd64_src.go

```diff
@@ -0,0 +1,148 @@
+// +build ignore
+
+// AVX2 fourway parallelized KeccaK-f[1600].
+
+package main
+
+import (
+	. "github.com/mmcloughlin/avo/build"
+	. "github.com/mmcloughlin/avo/operand"
+)
+
+func main() {
+	ConstraintExpr("amd64")
+
+	TEXT("f1600x4AVX2", 0, "func(state *[100]uint64, rc *[24]uint64)")
+
+	// Pragma("noescape")  // XXX
+	// We want state to be 16 byte alligned.  Go only guarantees an
+	// (array of) uint64 to be alligned on 8 bytes except if it's heap
+	// allocated.  Thus we do not add the noescape pragma to ensure that
+	// state is heap allocated and thus 16 byte alligned.
```

Comment:
Hmm, this feels fragile. Do you need alignment for `VMOVDQA`, or is there another reason? Have you tried `VMOVDQU`? I suspect the unaligned version will incur a small or negligible performance hit, probably worth it for correctness.

Reply:
Do you know how to request the go compiler for aligned memory ?

Reply:
@mmcloughlin Memory is not just loaded at the start, but it’s read and loaded between each round. I think it’ll have a measurable impact. (Vector operations which do allow unaligned pointers are generally still faster with aligned pointers.)

@armfazh I don’t think there is good support for it yet.

Reply:
@bwesterb I made the change to `VMOVDQU` locally and benchmarked it each time.

```
$ cat aligned.txt unaligned.txt 
goos: linux
goarch: amd64
pkg: github.com/cloudflare/circl/internal/shakex4
BenchmarkF1600x4-8   	 2147116	       556 ns/op
PASS
ok  	github.com/cloudflare/circl/internal/shakex4	1.762s
goos: linux
goarch: amd64
pkg: github.com/cloudflare/circl/internal/shakex4
BenchmarkF1600x4-8   	 2194231	       546 ns/op
PASS
ok  	github.com/cloudflare/circl/internal/shakex4	1.754s
```

Unaligned was faster in this quick uncontrolled test, which is probably noise, but that just goes to show how small the effect is. I encourage you to benchmark this yourself, and I think you'll find the performance change (if any at all) is so small that it's not worth the trouble of forcing Go to give you aligned memory, since as you say there is not good support for it.

https://lemire.me/blog/2012/05/31/data-alignment-for-speed-myth-or-reality/

I'm happy to be proven wrong, but I just think you should use the unaligned access until you can demonstrate alignment really helps.

Reply:
The distinction in time is not observed in our current processors and laptops, say after haswell u-arch (2014 approx). 
But in it will appear in older processors. In those one probably see a difference in performance.

Reply:
For the matters of this PR and CIRCL, I am ok with using unaligned memory access, until Go provides a way to ensure this.

Reply:
@armfazh AVX2 was introduced in Haswell. This code can't run on older microarchitectures anyway.

Reply:
256-bit registers appear in the Sandy Bridge arch and it was called AVX. Hence, there was load aligned and unaligned before haswell.

Reply:
Yes, but for the purposes of this PR, that doesn't matter. Other instructions in this code require AVX2.

Reply:
I ran my own benchmarks.  There doesn't seem to be a big difference between VMOVDQA and VMOVDQU for **aligned** memory.  If I change alignment, I get a noticeable slowdown.  (I ran the benchmarks interleaved and then sorted them to account for my CPU throttling.)  

```
BenchmarkF1600x4-8               2783442           428 ns/op                     
BenchmarkF1600x4-8               2850756           421 ns/op                     
BenchmarkF1600x4-8               2858080           431 ns/op                     
BenchmarkF1600x4aligned-8        2700562           430 ns/op                     
BenchmarkF1600x4aligned-8        2888199           437 ns/op                     
BenchmarkF1600x4aligned-8        2889950           425 ns/op                     
BenchmarkF1600x4plusEight-8      2475714           487 ns/op                     
BenchmarkF1600x4plusEight-8      2558049           455 ns/op                     
BenchmarkF1600x4plusEight-8      2564606           465 ns/op                     
BenchmarkF1600x4plusFour-8       2580600           463 ns/op                     
BenchmarkF1600x4plusFour-8       2585941           475 ns/op                     
BenchmarkF1600x4plusFour-8       2585942           467 ns/op                     
BenchmarkF1600x4plusOne-8        2505846           477 ns/op                     
BenchmarkF1600x4plusOne-8        2581089           479 ns/op                     
BenchmarkF1600x4plusOne-8        2647069           463 ns/op                     
BenchmarkF1600x4plusTwo-8        2512466           474 ns/op                     
BenchmarkF1600x4plusTwo-8        2519284           470 ns/op                     
BenchmarkF1600x4plusTwo-8        2531073           478 ns/op
```

Reply:
Oh awesome, interesting results.

I would still recommend switching to `VMOVQU`, but keep the no-`noescape` technique you are using to get aligned memory from Go. You'll get the performance gain, but it will continue to work if you don't get aligned memory _for whatever reason_.

Reply:
The no-noescape caused quite a bit of allocator overhead, so instead I opted to allocate an `[103]uint64` and move my offset to ensure 32 byte alignment.